vnext/ReactUWP/Modules/NativeUIManager.cpp

```diff
@@ -1011,14 +1011,14 @@ void NativeUIManager::UpdateExtraLayout(int64_t tag) {
   if (shadowNode == nullptr)
     return;
 
-  if (shadowNode->IsExternalLayoutDirty()) {
-    YGNodeRef yogaNode = GetYogaNode(tag);
-    if (yogaNode)
-      shadowNode->DoExtraLayoutPrep(yogaNode);
+  auto control = shadowNode->GetView().try_as<winrt::Control>();
```

Comment:
There are a couple issues here:
1) It's running UpdateLayout on controls every time we run layout.  This is more expensive than it needs to be.  We really should only UpdateLayout just after the control was created.
2) If a ViewManager creates a tree of XAML elements to back the ShadowNode (there may be a hierarchy like in View), and if it creates a control somewhere in the children part of the backing view, this code won't detect that case.

To solve #1 we could store away the appropriate controls in a vector that we populate in NativeUIManager::CreateView.
To solve #2 we could add a virtual function to ShadowNodeBase that each ShadowNode can override.  The function returns true if it creates a control somewhere in its hierarchy (or any other case we find that needs forceful layout).  We would use this information to know which views need UpdateLayout.  The default would be false so we only need to update a few ShadowNodes to return true.

Reply:
So just to confirm, in DoLayout I would be iterating through the vector generated in CreateView and checking the virtual function (like needsForceLayout) and then if that returns true then I'm running the layout on it? I think I'm a little confused about the second part, I'm not sure how to go about detecting that case from NativeUIManager, or how the shadownode would override the 'needsForceLayout' function.

Reply:
Yes to your first question.
For the second part:
-Add a virtual function to ShadowNodeBase, which returns false, and override it to return true in the specific types of ShadowNodes that need to force layout
-Have your vector store ShadowNode tags (of newly created views)
-In your loop, lookup the ShadowNode from its tag and call the virtual function.  If it returns true then grab the ShadowNode's backing XamlView and run UpdateLayout on it.